### ISSUE TITLE
Gvim spellcheck right-click change not consistently working

### ISSUE THREAD

**Issue by Erotemic** [vim/vim#1483]
I have a latex document and in it I've spelled the word "intermittently" incorrectly as intermittenly. 

```tex
    \documentclass[10pt,twocolumn,letterpaper]{article}
    \usepackage[T1]{fontenc}
    
    \begin{document}
    
    \title{A minimum working example of a bug}
    
    The bug is that intermittenly is spelled wrong, and has a yellow squiggle.
    However, if I right click and select ``Change "intermittenly" to
      intermittently'' it usually does nothing, but sometimes it will change it.
    \end{document}
```

Spell check correctly identifies the two instances of this with a yellow squiggle underneath. However, if I right click to fix it does not always work. I can't figure out the conditions to consistently reproduce the error. 

Typically nothing happens when I click change. However, sometimes it works and changes the word. Furthermore, if I undo the change and then right click and change again it typically does nothing again. 

I'm at a loss to what is causing this behavior. 

I've tested this with a very minimal .vimrc

```vim
    source $VIMRUNTIME/mswin.vim
    behave mswin
```

**Comment by vim-ml** [comment]
On Thursday, February 16, 2017 at 10:34:49 AM UTC-6, Jon Crall wrote:
> I have a latex document and in it I've spelled the word "intermittently" incorrectly as intermittenly.
> 
>     \documentclass[10pt,twocolumn,letterpaper]{article}
>     \usepackage[T1]{fontenc}
>     
>     \begin{document}
>     
>     \title{A minimum working example of a bug}
>     
>     The bug is that intermittenly is spelled wrong, and has a yellow squiggle.
>     However, if I right click and select ``Change "intermittenly" to
>       intermittently'' it usually does nothing, but sometimes it will change it.
>     \end{document}
> 
> Spell check correctly identifies the two instances of this with a yellow squiggle underneath. However, if I right click to fix it does not always work. I can't figure out the conditions to consistently reproduce the error.
> 
> Typically nothing happens when I click change. However, sometimes it works and changes the word. Furthermore, if I undo the change and then right click and change again it typically does nothing again.
> 
> I'm at a loss to what is causing this behavior.
> 
> I've tested this with a very minimal .vimrc
> 
>     source $VIMRUNTIME/mswin.vim
>     behave mswin
> 

I don't know what could be causing your right-click error, but in case you just need a fast solution, you can use z= on your keyboard to get spelling suggestions for misspelled words.

**Comment by nuko8** [comment]
If I put the cursor anywhere on a misspelled word before giving it a right click, then a menu with a list of correct spells popped up immediately after a right click was given.  I made sure that the Athena, Motif, GTK2 and GTK3 GUIs behaved like that with the TeX example above.

**Comment by Erotemic** [comment]
The menu with correct words appeared, its just that choosing one of these words does not replace the incorrect word all of the time.

**Comment by nuko8** [comment]
I wanted to mean by the previous comment that word replacement also worked fine.  Sorry about the obscurity.  Then, I can't reproduce the issue, and somehow, your description of the issue changed from "does not always work" to "does not replace the incorrect word all of the time", I can't seem to understand the issue well.

**Comment by Erotemic** [comment]
My original issue says it does not work consistently. Sometimes it does, typically it doesn't. 

I just tested it on a different computer and the behavior still happens.

**Comment by chrisbra** [comment]
Hm, I haven't used that feature ever before. From my quick testing, the only problem I see is that this actually only works, if you first position the cursor on the correct word, since on a rightclick the cursor does not move. So if you have this buffer:
1 worng
2 rigt
with the cursor is on the word  rigt on the second line and you right click on the word of the first line,
it will show you suggestions for the word 'rigt' (and change the word on the second line, rather the word on the first line). So that is at least unexpected. Perhaps on a right click, the cursor should be positioned first. Also, you might want to try to debug the SpellPopup() function. So do a `:breakadd func *SpellPopup` and right click on a word. Read the help at `:h debug-scripts` to know how to step through the function.

**Comment by chrisbra** [comment]
Also you might want to check if this happens, without any plugins: `vim -u NONE -N -g --cmd 'syntax on' --cmd 'set mousemodel=popup spell spelllang=en'`  Perhaps something is interfering with the default setup. (it might also help to post an easily reproducible example.

**Comment by Erotemic** [comment]
I tried using `vim -u NONE -N -g --cmd 'syntax on' --cmd 'set mousemodel=popup spell spelllang=en'` to run without any plugins, and I experienced the same issue. The first two times that I attempted to change the spelling nothing happened, but on the third time it worked. I also found that `z=` always works.

**Comment by vim-ml** [comment]
On Thursday, February 23, 2017 at 12:53:11 PM UTC-4, Jon Crall wrote:
> I tried using vim -u NONE -N -g --cmd 'syntax on' --cmd 'set mousemodel=popup spell spelllang=en' to run without any plugins, and I experienced the same issue. The first two times that I attempted to change the spelling nothing happened, but on the third time it worked. I also found that z= always works.

Try 
set mousemodel=popup_setpos

**Comment by Erotemic** [comment]
Same behavior.

**Comment by nuko8** [comment]
@Erotemic 

In addition to @chrisbra 's suggesions (Have you tried the SpellPopup function debug?  If you already have a result and could tell us about it, that would be helpful), would you try the procedure below with your gvim?  That could help us guess what happens to you with it.

1. As suggested by @chrisbra , do `vim -u NONE -N -g --cmd 'syntax on' --cmd 'set mousemodel=popup spell spelllang=en'`
2. Open the TeX file you mentioned in the initial post.
3. Place the cursor over one of the misspelled words.
4. Right-click.  You'll have a popup menu having the submenu `change "intermittenly" to`.
5. Move the pointer into the submenu, and stop moving it there.  When the pointer goes into the submenu, you'll see a list of candidates popping up, but DON'T go into the list.
6. Click once.  This will close the candidate list.
7. Click once again.  This will reopen it.
8. Now go into the list: Move the cursor to the correctly spelled word, and click or right-click when you find the word is selected.

Then, has the misspelled word been replaced with the correct word for you, or not?

**Comment by Erotemic** [comment]
Yes, it works. By clicking on the submenu and then clicking again, the word is always replaced. 

Does this give you an idea to what is happening?

I'm not sure what "SpellPopup function debug" is, or how to use it.  Did I run this when I followed the click procedure you outlined?

**Comment by nuko8** [comment]
> Yes, it works. By clicking on the submenu and then clicking again, the word is always replaced.

OK, now you know how to get it to always work, though it's surely clunky.  Seems I need more info to see if I can do more for you. So can you try this?: click in place to choose the word, right-click there (then the menu pops up), move the pointer to the label `change "intermittenly" to` on the menu (then the submenu of correct spellings pops up as the pointer goes into it), keep the pointer going across the label and go straight to another label `intermittently` on the submenu, and stop there.  Then, right-click.  Does this work for you to correct the word?

In addition, I want to know which gvim you are using.  Athena, Motif, GTK2 or GTK3?

> I'm not sure what "SpellPopup function debug" is, or how to use it. 

That's what is written at the bottom of @chrisbra 's first message:
>Also, you might want to try to debug the SpellPopup() function. So do a :breakadd func *SpellPopup and right click on a word. Read the help at :h debug-scripts to know how to step through the function.

**Comment by Erotemic** [comment]
Yes, right clicking the submenu item seems to work as well. Its just the left click that behaves oddly. 


> In addition, I want to know which gvim you are using. Athena, Motif, GTK2 or GTK3?

It seems to be GTK2. Here is the output of vim --version

```
(venv2) joncrall@hyrule:~/code/drawtree$ vim --version
VIM - Vi IMproved 8.0 (2016 Sep 12, compiled Feb  7 2017 12:27:58)
Included patches: 1-314
Compiled by joncrall@hyrule
Huge version with GTK2 GUI.  Features included (+) or not (-):
+acl             +file_in_path    +mouse_sgr       +tag_old_static
+arabic          +find_in_path    -mouse_sysmouse  -tag_any_white
+autocmd         +float           +mouse_urxvt     -tcl
+balloon_eval    +folding         +mouse_xterm     +termguicolors
+browse          -footer          +multi_byte      +terminfo
++builtin_terms  +fork()          +multi_lang      +termresponse
+byte_offset     +gettext         -mzscheme        +textobjects
+channel         -hangul_input    +netbeans_intg   +timers
+cindent         +iconv           +num64           +title
+clientserver    +insert_expand   +packages        +toolbar
+clipboard       +job             +path_extra      +user_commands
+cmdline_compl   +jumplist        -perl            +vertsplit
+cmdline_hist    +keymap          +persistent_undo +virtualedit
+cmdline_info    +lambda          +postscript      +visual
+comments        +langmap         +printer         +visualextra
+conceal         +libcall         +profile         +viminfo
+cryptv          +linebreak       -python          +vreplace
+cscope          +lispindent      +python3         +wildignore
+cursorbind      +listcmds        +quickfix        +wildmenu
+cursorshape     +localmap        +reltime         +windows
+dialog_con_gui  -lua             +rightleft       +writebackup
+diff            +menu            -ruby            +X11
+digraphs        +mksession       +scrollbind      -xfontset
+dnd             +modify_fname    +signs           +xim
-ebcdic          +mouse           +smartindent     +xpm
+emacs_tags      +mouseshape      +startuptime     +xsmp_interact
+eval            +mouse_dec       +statusline      +xterm_clipboard
+ex_extra        +mouse_gpm       -sun_workshop    -xterm_save
+extra_search    -mouse_jsbterm   +syntax          
+farsi           +mouse_netterm   +tag_binary      
   system vimrc file: "$VIM/vimrc"
     user vimrc file: "$HOME/.vimrc"
 2nd user vimrc file: "~/.vim/vimrc"
      user exrc file: "$HOME/.exrc"
  system gvimrc file: "$VIM/gvimrc"
    user gvimrc file: "$HOME/.gvimrc"
2nd user gvimrc file: "~/.vim/gvimrc"
       defaults file: "$VIMRUNTIME/defaults.vim"
    system menu file: "$VIMRUNTIME/menu.vim"
  fall-back for $VIM: "/home/joncrall/share/vim"
Compilation: gcc -c -I. -Iproto -DHAVE_CONFIG_H -DFEAT_GUI_GTK  -pthread -I/usr/include/gtk-2.0 -I/usr/lib/x86_64-linux-gnu/gtk-2.0/include -I/usr/include/gio-unix-2.0/ -I/usr/include/cairo -I/usr/include/pango-1.0 -I/usr/include/atk-1.0 -I/usr/include/cairo -I/usr/include/pixman-1 -I/usr/include/libpng12 -I/usr/include/gdk-pixbuf-2.0 -I/usr/include/libpng12 -I/usr/include/pango-1.0 -I/usr/include/harfbuzz -I/usr/include/pango-1.0 -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -I/usr/include/freetype2   -g -O2 -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1       
Linking: gcc   -L/usr/local/lib -Wl,--as-needed -o vim   -lgtk-x11-2.0 -lgdk-x11-2.0 -lpangocairo-1.0 -latk-1.0 -lcairo -lgdk_pixbuf-2.0 -lgio-2.0 -lpangoft2-1.0 -lpango-1.0 -lgobject-2.0 -lglib-2.0 -lfontconfig -lfreetype -lSM -lICE -lXpm -lXt -lX11 -lXdmcp -lSM -lICE  -lm -ltinfo -lnsl  -lselinux  -lacl -lattr -lgpm -ldl     -L/usr/lib/python3.5/config-3.5m-x86_64-linux-gnu -lpython3.5m -lpthread -ldl -lutil -lm      
```

> That's what is written at the bottom of @chrisbra 's first message:

I must have missed that message. I also noticed the behavior where you actually have to click on a word before you can right click and change its spelling, but that seems like a more minor issuel. I played around with debugging SpellPopup a little, but it looks like I'm going to need to do some reading before I can make any use of what it tells me.

**Comment by nuko8** [comment]
> Yes, right clicking the submenu item seems to work as well. Its just the left click that behaves oddly.

So, with reservation about that oddity, can we say the spellcheck just works?

As for the Athena, Motif and GTK3 GUIs, both a left- and a right-click will do for selecting a word.

> It seems to be GTK2. Here is the output of vim --version

...And this is at the core of the issue. 

The GTK2 gvim implements popup menus using `gtk_menu_popup()`, and in `gtk+-2.24.31/gtk/gtkmenu.c`, we have the following comment which hints both what you've experienced with the gvim so far and how I was able to guess that "left-click on menu and then left-click on submenu" and "right-click on submenu" might work for you:
```
  /* We want to receive events generated when we map the menu; unfortunately,
   * since there is probably already an implicit grab in place from the
   * button that the user used to pop up the menu, we won't receive then --
   * in particular, the EnterNotify when the menu pops up under the pointer.
   *
   * If we are grabbing on a parent menu shell, no problem; just grab on
   * that menu shell first before popping up the window with owner_events = TRUE.
   *
   * When grabbing on the menu itself, things get more convuluted - we
   * we do an explicit grab on a specially created window with
   * owner_events = TRUE, which we override further down with a grab
   * on the menu. (We can't grab on the menu until it is mapped; we
   * probably could just leave the grab on the other window, with a
   * little reorganization of the code in gtkmenu*).
   */
```
In short, it's a GTK2 bug.  Although we still sometimes have a GTK2 update for bug fixes, it is not developed these days.  In addition, the `gtk_menu_popup()` function has been deprecated since GTK+ 3.22.  Accordingly, it is very hard for us to tell when the bug is fixed (Personally, I think it won't be fixed).

**Comment by chrisbra** [comment]
so please check with a gtk3 gui then.

**Comment by Erotemic** [comment]
I recompiled with GTK3 and unfortunately the behavior persists. (But using right click still works).

I seem to be on gtk 3.18.9, which means maybe `gtk_menu_popup` is still causing issues. Perhaps I should try and upgrade my libgtk files to 3.22?

I did recompile with Athena and was able to verify that the unexpected behavior does not exist.

**Comment by nuko8** [comment]
> I seem to be on gtk 3.18.9, which means maybe gtk_menu_popup is still causing issues. 

Yes.

> Perhaps I should try and upgrade my libgtk files to 3.22?

I think it's a good idea if it meets your whole needs.  

While it has deprecated the function, 3.22 still has gtk_menu_popup(), but the function is implemented with a single call of a new API function.  The GTK3 gvim uses new API for 3.22.2.  3.22.4 or later is preferable.

**Comment by Erotemic** [comment]
Thanks for all the help. I love understanding more about how vim works under the hood. I'll upgrade to 3.22 as soon as possible, but for now I think this is sufficient.
